File: 799.champagne-tower/solution.kt
class Solution {
    fun champagneTower(poured: Int, query_row: Int, query_glass: Int): Double {
        var currentRow = DoubleArray(1) { poured.toDouble() }
        repeat(query_row) {
            val nextRow = DoubleArray(currentRow.size + 1)
            for (c in 0 until nextRow.lastIndex) {
                val q = (currentRow[c] - 1) / 2
                if (q > 0) {
                    nextRow[c] += q
                    nextRow[c + 1] += q
                }
            }
            currentRow = nextRow
        }
        return minOf(1.0, currentRow[query_glass])
    }
}

File: 799.champagne-tower/solution.go
func champagneTower(poured int, query_row int, query_glass int) float64 {
    current := []float64{float64(poured)}
    for r := range query_row {
        next := make([]float64, r + 2)
        for c := range r + 1 {
            if q := (current[c] - 1) / 2; q > 0 {
                next[c] += q
                next[c + 1] += q
            }
        }
        current = next
    }
    return min(1.0, current[query_glass])
}